crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,321 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::bigtable::proto::bigtable::v2::bigtable_client::BigtableClient as BigtableInternalClient;
+use crate::bigtable::proto::bigtable::v2::mutate_rows_request::Entry;
+use crate::bigtable::proto::bigtable::v2::mutation::SetCell;
+use crate::bigtable::proto::bigtable::v2::read_rows_response::cell_chunk::RowStatus;
+use crate::bigtable::proto::bigtable::v2::{
+    mutation, MutateRowsRequest, MutateRowsResponse, Mutation, ReadRowsRequest, RowSet,
+};
+use crate::{Checkpoint, KeyValueStore, TransactionData};
+use anyhow::Result;
+use async_trait::async_trait;
+use gcp_auth::{Token, TokenProvider};
+use http::{HeaderValue, Request, Response};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+use std::task::{Context, Poll};
+use std::time::Duration;
+use sui_types::base_types::{SequenceNumber, TransactionDigest};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use tonic::body::BoxBody;
+use tonic::codegen::Service;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Streaming;
+
+const COLUMN_FAMILY_NAME: &str = "sui";
+const COLUMN_QUALIFIER: &str = "";
+const OBJECTS_TABLE: &str = "objects";
+const CHECKPOINTS_TABLE: &str = "checkpoints";
+const TRANSACTIONS_TABLE: &str = "transactions";
```

Comment:
I think we'll also want a watermark table that we can use to track the height of each of these tables.

Also it looks like we're writing 100% of the data into a single column, can we break the data up into multiple columns per table, where it makes sense of course. Specifically for Checkpoints and Transactions

Reply:
I don't see a need for a watermark table, specially having separate watermarks(single one works ok for existing KV store). In general, it’s the role of the progress store in the ingestion framework. 
It’s mocked in this PR, but initially, we'll use the existing progress store that we already utilize for various pipelines.

Also, with Bigtable, you don't technically need a dedicated progress store; the largest key in the checkpoints table is the watermark

Reply:
> we're writing 100% of the data into a single column

Yeah, this PR is very basic; the client doesn't support multiple columns yet. 
It makes sense to split data into multiple columns (DDB is already split). 
But I want to talk through what that implies. Are we changing the API by allowing subqueries for data? Do we want to enforce a similar structure/partition for all other potential KV backends?

Reply:
> I don't see a need for a watermark table, specially having separate watermarks(single one works ok for existing KV store).

(Separate) watermarks are used by GraphQL's consistent read logic

Reply:
maybe I'm missing something. @amnn: can you elaborate on how this is relevant here? There’s a single ingestion pipeline. The last processed checkpoint can be derived without a separate watermark table and indicates what data is present in the KV store. All data is immutable

Reply:
GraphQL advertises an "available range" of checkpoints based on the checkpoint upper- and lowerbounds of all its data sources, this makes sure that users can't issue queries that we will only have partial data for (e.g. if our object filter tables have data up to checkpoint `C` but the KV store only has data up to checkpoint `D < C` then we artificially hold back what we consider to be the latest checkpoint, to `D`, so that you can't return a result that says some object exists but we don't have its contents).

This means that in future, when GraphQL is reading from the KV store, it needs to know what the max checkpoint is of data in the KV store. If there's some way to read that off the KV store, then that's all good (as long as it's a lagging indicator -- i.e. this source can't advertise that there is data available in the KV store before it's accessible from there).

Reply:
Yeah, it would be good to have an efficient way to read the latest available checkpoint. Does this mean that we would need to download the whole checkpoint every time though? That seems not ideal. It's workable if it's cheap to query some small prefix of the latest checkpoint row efficiently to figure out what the latest checkpoint is, but otherwise it seems like it would be best to add the explicit watermark.

Reply:
we don't need to download the entire row, just a key without any columns. But I may need to follow up with you regarding the `every time` part, as doing that for every query seems a bit off